sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -221,23 +234,53 @@
       Coder<InputT> inputCoder;
       WindowingStrategy<InputT, ?> windowingStrategy;
 
+      ImmutableMap.Builder<TupleTag<?>, SideInputSpec> tagToSideInputSpecMap =
+          ImmutableMap.builder();
+      ParDoPayload parDoPayload;
       try {
         RehydratedComponents rehydratedComponents = RehydratedComponents.forComponents(
             RunnerApi.Components.newBuilder()
                 .putAllCoders(coders).putAllWindowingStrategies(windowingStrategies).build());
-        ParDoPayload parDoPayload = ParDoPayload.parseFrom(pTransform.getSpec().getPayload());
-        if (parDoPayload.getSideInputsCount() != 0) {
-          throw new UnsupportedOperationException("Side inputs not yet supported.");
-        }
+        parDoPayload = ParDoPayload.parseFrom(pTransform.getSpec().getPayload());
         doFn = (DoFn) ParDoTranslation.getDoFn(parDoPayload);
         mainOutputTag = (TupleTag) ParDoTranslation.getMainOutputTag(parDoPayload);
-        // There will only be one due to the check above.
-        RunnerApi.PCollection mainInput = pCollections.get(
-            Iterables.getOnlyElement(pTransform.getInputsMap().values()));
+        String mainInputTag = Iterables.getOnlyElement(Sets.difference(
+            pTransform.getInputsMap().keySet(), parDoPayload.getSideInputsMap().keySet()));
+        RunnerApi.PCollection mainInput =
+            pCollections.get(pTransform.getInputsOrThrow(mainInputTag));
         inputCoder = (Coder<InputT>) rehydratedComponents.getCoder(
             mainInput.getCoderId());
         windowingStrategy = (WindowingStrategy) rehydratedComponents.getWindowingStrategy(
             mainInput.getWindowingStrategyId());
+
+        // Build the map from tag id to side input specification
+        for (Map.Entry<String, RunnerApi.SideInput> entry
+            : parDoPayload.getSideInputsMap().entrySet()) {
+          String sideInputTag = entry.getKey();
+          RunnerApi.SideInput sideInput = entry.getValue();
+          checkArgument(
+              Materializations.MULTIMAP_MATERIALIZATION_URN.equals(
+                  sideInput.getAccessPattern().getUrn()),
+              "This runner is only capable of dealing with %s materializations "
```

Comment:
<!--new_thread; commit:c3d73c64facc4d4670f203a8fda5fafa1fb8ef8b; resolved:0-->
s/This runner/This SDK/

Reply:
Done